### ISSUE TITLE
Reconnection failed when shutdown rabbitMq on server and swith to another cluster rabbit server

### ISSUE THREAD

**Issue by Fbernardpro** [guidesmiths/rascal#82]
Hi,
i would like to create a nodejs script with rascal api, which is listening a rabbit queue and automatiquely reconnect to a RabbitMq server after a Rabbit shutdown and switch to another server (cluster).
please find below my code in attachment,  based on the Rascal documentations.
My configuration : rabbit server cluster of 2 servers
rabbitMq 3.7.5
javascript nodejs with rascal api : 4.7.0

[index.js.txt](https://github.com/guidesmiths/rascal/files/4343032/index.js.txt)
[configRascal.js.txt](https://github.com/guidesmiths/rascal/files/4343035/configRascal.js.txt)

It runs fine when i start de project , but , when i test the shutdown i have this message :

```
[2020-03-11T17:13:28.941Z] DEBUG: psf-ms-rascal-1.0.0/15684 on SCU48652 (L:\psf-ms-rascal\src\amq\listeners\index.js:15 in module.exports): Rascal listener initialized --------------
[2020-03-11T17:13:28.943Z] DEBUG: psf-ms-rascal-1.0.0/15684 on SCU48652 (L:\psf-ms-rascal\src\amq\listeners\index.js:19 in module.exports): Rascal listener running fine --------------
events.js:183
throw er; // Unhandled 'error' event

Error: read ECONNRESET
at _errnoException (util.js:1024:11)
at TCP.onread (net.js:615:25)
[nodemon] app crashed - waiting for file changes before starting...
```

It's normal.

But when i start rabbitmq again , it doesn't reconnect automatiquely.

I dont know how i can fix simply this issue.

this issue could be similar as one of them , but the code looked is completely different. So i need your.
Thanks a lot .
Fbernardpro.

**Comment by cressie176** [comment]
Hi @Fbernardpro,

If I understand correctly this is your process...

1. Start application using nodemon
1. Stop RabbitMQ
1. Instead of handling the connection error, crash the application
1. Restart RabbitMQ
1. Wait for nodemon to restart the application based on file changes

The problem with this approach, is that by step 4 the application (including the Rascal code) isn't running, so there's nothing it can do to recover. If you want to keep going with this approach you will need to see if nodemon can be configured to restart the application even if files haven't been changed.

If instead you want to use Rascal's error handling, then you'll need to be sure to listen to the error handlers correctly. If you do this then the application will not throw an Unhandled Error. The code you uploaded seems to have padded some of the event names with spaces. Potentially this could be your problem. I would start by getting one of the examples running locally without changes (or possibly [this](https://github.com/cressie176/amqplib-express-example), then adapting them. You can also enable Rascal's debug for more details by setting the `DEBUG` environment variable to `rascal:*`.

**Comment by Fbernardpro** [comment]
Hi cressie176,

thank you  for answer. 
Your understanding is exactly what i would like to implement.
But as you can see , my code is very simple and i took it on the Rascal npm web site.

In fact , we manage a cluster with differents Rabbit servers allowed.
When a server is unavailable,  my Rascal api must switch to a second available server.
In my test environnement , a don'have a cluster , only one.
To simulate this state, i shutdown rabbitMq , i restart it again.

As I'm a beginner in this kind of api and i need some examples managing the RabbitMQ.
So I will study your codes that you provided me and follow your advices. 
I let you know my progress.
Thank you again.

Best Regards.

**Comment by cressie176** [comment]
👍 Just in case you haven't found them, [these](https://github.com/guidesmiths/rascal/tree/master/examples) are the examples I was referring to.

**Comment by Fbernardpro** [comment]
Ok, that's fine. Thank you Stephen.

Envoyé depuis Yahoo Mail pour Android 
 
  Le mar., mars 17, 2020 à 21:08, Stephen Cresswell<notifications@github.com> a écrit :   


 👍 Just in case you haven't found them, these are the examples I was referring to.

**Comment by Fbernardpro** [comment]
Hi Stephen,

if i understand i have to add an handler on broker.create.
After looking examples , there are some handler on queue messages under broker.on.
But i think for my understand, it's under broker.create ?
What do you think about my example in attachement.
THe problem is how to restart it on handler ?
Another example give an automatically  connection but with Rabbit only.  
[index_bis.txt](https://github.com/guidesmiths/rascal/files/4355329/index_bis.txt)
I would be greatfully if i could help me.

Sorry I can't test it today, because  i'm working at home now and for some days. and i have no rabbitmq available.
 
 Best Regards.
Francois

**Comment by cressie176** [comment]
Hi Francois,

The best advice I can give is to start by getting the examples working against a local instance of RabbitMQ using something like [docker](https://hub.docker.com/_/rabbitmq). Then you can bounce RabbitMQ locally and see the behaviour.

Once you have this working, you can adapt the example and it should continue to work.

**Comment by Fbernardpro** [comment]
Hi Stephen,

i saw that i can install a rabbitmq on my pc with erlang as required before. I will do that if it's free.
And test again.
And about my script index_bis ? is it the right way ?
Thank's a lot
Best Regards.
Francois

**Comment by cressie176** [comment]
Hi Francois,

I appreciate this is all a bit new, and possibly some of the advice I'm giving you seems harder than you think. Using [docker](https://www.docker.com/products/docker-desktop) is likely to be much easier than trying to install RabbitMQ, and will probably save you a huge amount of hassle. Plus if you need further help it will make things easier you can run commands like [these](https://github.com/cressie176/amqplib-express-example#testing-automated-recovery-connection-close). RabbitMQ and Docker are both open source.

I did look at your script, but it's not how I would have coded things. There's no "right" way, but the approach I recommend is to follow the examples.

**Comment by Fbernardpro** [comment]
ok , thank's again for advices.
I let you know my progress.
Regards.
Francois

**Comment by cressie176** [comment]
I'm going to close this, as it's not a technical issue and I don't think there's much more I can do to help.

**Comment by Fbernardpro** [comment]
Hi Stephen,
Sorry, I didn't have any time to continue to investigate on rabbit my issue these time.
I will continue later after our Containment in France. Complicate to work in an abnormal state.
I will let you know in time.May be reopen ir close definitevely this issue.
Thank you again.Regards.Francois.
Envoyé depuis Yahoo Mail pour Android 
 
  Le mar., avr. 7, 2020 à 22:47, Stephen Cresswell<notifications@github.com> a écrit :   


Closed #82.
